### PR TITLE
Add MelGAN compatible models.

### PR DESCRIPTION
**Pull request by cfrancesco**
Audio pre-processing is now enclosed in a class. Normalization has it's own logic and class within the pre-processing logic.

### CODE REVIEW

File: config/wavernn/forward_config.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+debug: False
+
+# LOGGING
+validation_frequency: 1_000
+prediction_frequency: 1_000
+weights_save_frequency: 5_000
+train_images_plotting_frequency: 1_000
+keep_n_weights: 5
+keep_checkpoint_every_n_hours: 12
+n_steps_avg_losses: [100, 500, 1_000, 5_000]
```
Comment (cschaefer26) [inline config/wavernn/forward_config.yaml:41]: Add a comment explaining what n_steps_avg_losses is.

File: docs/index.md [hunk 2]
```diff
@@ -8,33 +8,42 @@
...
 <p class="text">President Trump met with other leaders at the Group of twenty conference.</p>
 
-| forward | autoregressive |
+| forward + wavernn | autoregressive + wavernn |
 |:---:|:---:|
 |<audio src="https://github.com/as-ideas/tts_model_outputs/blob/master/ljspeech_forward_transformer/trump.wav?raw=true" controls preload></audio>|<audio src="https://github.com/as-ideas/tts_model_outputs/blob/master/ljspeech_transformertts/Trump.wav?raw=true" controls preload></audio>|
+| <b>forward + melgan</b> | <b>autoregressive + melgan</b> |
+|<audio src="https://github.com/as-ideas/tts_model_outputs/blob/master/ljspeech_transformerTTS_melgan/trump_forward_melgan.wav?raw=true" controls preload></audio>|<audio src="https://github.com/as-ideas/tts_model_outputs/blob/master/ljspeech_transformerTTS_melgan/trump_autoregressive_melgan.wav?raw=true" controls preload></audio>|
 
 <p class="text">Scientists, at the CERN laboratory, say they have discovered a new particle.</p>
```
Comment (cschaefer26) [inline docs/index.md:24]: I still believe that there should be no comma here :-)

File: utils/audio.py [hunk 3]
```diff
@@ -1,61 +1,103 @@
...
-    return wav
+class Audio():
+    def __init__(self, config: dict):
+        self.config = config
+        self.normalizer = getattr(sys.modules[__name__], config['normalizer'])(config)
+    
+    def _normalize(self, S):
+        return self.normalizer.normalize(S)
+    
+    def _denormalize(self, S):
```
Comment (cschaefer26) [inline utils/audio.py:15]: Make (de-)normalization public, its used.

File: utils/logging.py [hunk 4]
```diff
@@ -118,7 +119,7 @@ def display_scalar(self, tag, scalar_value, plot_all=False):
     
     @ignore_exception
     def display_audio(self, tag, mel):
-        wav = reconstruct_waveform(tf.transpose(mel), self.config)
+        wav = self.audio.reconstruct_waveform(tf.transpose(mel))
```
Comment (cschaefer26) [inline utils/logging.py:122]: Don't nest operations, I'd rather have one line more.

**Comment by cfrancesco** [comment]
> I would suggest simplifying the notebooks further by having a synth method that does both tts and vocoding so that ppl can just try out different strings in a single step.

Not trivial, due to import conflicts

File: config/melgan/data_config.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+# AUDIO
+sampling_rate: 22050
+n_fft: 1024
+mel_channels: 80
+hop_length: 256
+win_length: 1024
+f_min: 0
+f_max: 8000
+normalizer: MelGAN                 # which mel normalization to use from utils.audio.py
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply config/melgan/data_config.yaml:23]: The options are in the file that is commented, I'd say it's enough, especially in the prospect of an increasing list of options. Added the explicit options for now.

File: utils/audio.py (hunk 3 above)
Comment (cfrancesco) [reply utils/audio.py:15]: I did not like the structure of this anyways. Check the changes. Still private and used within this class.

File: README.md [hunk 6]
```diff
@@ -82,20 +91,21 @@ where `metadata.csv` has the following format:
...
 #### Training
 ```bash
-python train_autoregressive.py --config config/standard
+python train_autoregressive.py --config config/melgan
 ```
 ### Train Forward Model
 #### Compute alignment dataset
 First use the autoregressive model to create the durations dataset
 ```bash
-python extract_durations.py --config config/standard --binary --fix_jumps --fill_mode_next
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply README.md:98]: Where should I explain them, other than in the docstring?

[the remark this answers is not in the document]
Comment (cfrancesco) [reply README.md:98]: Added help strings

File: utils/audio.py [hunk 7]
```diff
@@ -1,61 +1,119 @@
...
+    
+    def mel_spectrogram(self, wav):
+        """ This is what the model is trained to reproduce. """
+        D = self._stft(wav)
+        S = self._linear_to_mel(np.abs(D))
+        return self._normalize(S)
+    
+    def reconstruct_waveform(self, mel, n_iter=32):
+        """ Uses Griffin-Lim phase reconstruction to convert from a normalized
+        mel spectrogram back into a waveform."""
```
Comment (datitran) [inline utils/audio.py:44]: Space missing in comments

File: utils/logging.py [hunk 8]
```diff
@@ -118,7 +119,7 @@ def display_scalar(self, tag, scalar_value, plot_all=False):
     
     @ignore_exception
     def display_audio(self, tag, mel):
-        wav = reconstruct_waveform(tf.transpose(mel), self.config)
```
Comment (cschaefer26) [inline utils/logging.py:121]: Do extra line instead of nesting.

File: notebooks/synthesize_autoregressive_wavernn.ipynb [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+        "colab_type": "code",
+        "outputId": "d1be75fc-b5b5-46b5-8f79-c898f2cdcac2",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 34
+        }
+      },
+      "source": [
+        "# Synthesize text\n",
+        "sentence = 'Scientists at the CERN laboratory, say that they have discovered a new particle.'\n",
```
Comment (cschaefer26) [inline notebooks/synthesize_autoregressive_wavernn.ipynb]: Still one comma too many :-)

**Pull request merged**